Userland/Libraries/LibGUI/Application.cpp

```diff
@@ -324,4 +325,73 @@ void Application::event(Core::Event& event)
     Object::event(event);
 }
 
+void Application::set_config_domain(String config_domain)
+{
+    m_config_domain = move(config_domain);
+}
+
+void Application::register_recent_file_actions(Badge<GUI::Menu>, Vector<NonnullRefPtr<GUI::Action>> actions)
+{
+    m_recent_file_actions = move(actions);
+    update_recent_file_actions();
+}
+
+void Application::update_recent_file_actions()
+{
+    size_t number_of_recently_open_files = 0;
+    auto update_action = [&](size_t index) {
+        auto& action = m_recent_file_actions[index];
+        char buffer = static_cast<char>('0' + index);
+        auto path = Config::read_string(m_config_domain, "RecentFiles"sv, MUST(String::from_utf8(StringView(&buffer, 1))));
+        if (path.is_empty()) {
+            action->set_visible(false);
+            action->set_enabled(false);
+        } else {
+            action->set_visible(true);
+            action->set_enabled(true);
+            action->set_text(path);
+            ++number_of_recently_open_files;
+        }
+    };
+    for (size_t i = 0; i < max_recently_open_files(); ++i)
+        update_action(i);
+
+    // Hide or show the "(No recently open files)" placeholder.
+    if (number_of_recently_open_files > 0) {
+        m_recent_file_actions.last()->set_visible(false);
+    } else {
+        m_recent_file_actions.last()->set_visible(true);
+    }
+}
+
+void Application::set_most_recently_open_file(String new_path)
+{
+    Vector<DeprecatedString> new_recent_files_list;
+
+    for (size_t i = 0; i < max_recently_open_files(); ++i) {
+        char buffer = static_cast<char>('0' + i);
+        auto key = StringView(&buffer, 1);
+        new_recent_files_list.append(Config::read_string(m_config_domain, "RecentFiles"sv, key));
+    }
+
+    new_recent_files_list.remove_all_matching([&](auto& existing_path) {
+        return existing_path.view() == new_path;
+    });
+
+    new_recent_files_list.prepend(new_path.to_deprecated_string());
+
+    for (size_t i = 0; i < max_recently_open_files(); ++i) {
+        auto& path = new_recent_files_list[i];
+        char buffer = static_cast<char>('0' + i);
+        auto key = StringView(&buffer, 1);
+        Config::write_string(
+            m_config_domain,
+            "RecentFiles"sv,
+            key,
+            path);
+    }
```

Comment:
Another option would be to concatenate the paths together and put them into a single config value. That would make it a bit simpler to watch the value for changes. (Just the single value to watch, no need to individually respond to one change notification per file.)

Though, it is a bit uglier in the .ini file to do that.

Reply:
I'll leave that question for our future selves to worry about. :)